libshpool/src/config.rs

```diff
@@ -12,36 +12,120 @@
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+    sync::{Arc, RwLock, RwLockReadGuard},
+};
 
 use anyhow::Context;
+use notify::Watcher;
 use serde_derive::Deserialize;
-use tracing::{info, instrument};
+use tracing::{info, warn};
 
 use super::{daemon::keybindings, user};
 
-#[instrument(skip_all)]
-pub fn read_config(config_file: &Option<String>) -> anyhow::Result<Config> {
-    let mut config = Config::default();
-    if let Some(config_path) = config_file {
-        info!("parsing explicitly passed in config ({})", config_path);
-        let config_str = fs::read_to_string(config_path).context("reading config toml (1)")?;
-        config = toml::from_str(&config_str).context("parsing config file (1)")?;
-    } else {
+/// Exposes the shpool config file, watching for file updates
+/// so that the user does not need to restart the daemon when
+/// they edit their config.
+///
+/// Users should never cache the config value directly and always
+/// access the config through the manager. The config may change
+/// at any time, so any cached config value could become stale.
+pub struct Manager {
+    /// The config value.
+    config: Arc<RwLock<Config>>,
+    watcher: Option<Arc<notify::RecommendedWatcher>>,
+}
+
+impl Manager {
+    // Create a new config manager.
+    pub fn new(config_file: Option<&str>) -> anyhow::Result<Self> {
         let user_info = user::info()?;
-        let mut config_path = PathBuf::from(user_info.home_dir);
-        config_path.push(".config");
-        config_path.push("shpool");
-        config_path.push("config.toml");
-        if config_path.exists() {
-            let config_str = fs::read_to_string(config_path).context("reading config toml (2)")?;
-            config = toml::from_str(&config_str).context("parsing config file (2)")?;
+        let mut default_config_path = PathBuf::from(user_info.home_dir);
+
+        let mut config = Config::default();
+        let config_path = if let Some(config_path) = config_file {
+            info!("parsing explicitly passed in config ({})", config_path);
+            let config_str = fs::read_to_string(config_path).context("reading config toml (1)")?;
+            config = toml::from_str(&config_str).context("parsing config file (1)")?;
+
+            Some(String::from(config_path))
+        } else {
+            default_config_path.push(".config");
+            default_config_path.push("shpool");
+            default_config_path.push("config.toml");
+            if default_config_path.exists() {
+                let config_str =
+                    fs::read_to_string(&default_config_path).context("reading config toml (2)")?;
+                config = toml::from_str(&config_str).context("parsing config file (2)")?;
+
+                default_config_path.clone().to_str().map(String::from)
+            } else {
+                None
+            }
+        };
+
+        let mut manager = Manager { config: Arc::new(RwLock::new(config)), watcher: None };
```

Comment:
nit: would be clearer to put this above the config_path assignment since this only uses config, not config_path.

Reply:
config get's reassigned in that block. I'm switching to a tuple assignment for both config and config_path at the same time to make it clearer.